packages/grid/x-data-grid-premium/src/hooks/features/clipboard/useGridClipboardImport.ts

```diff
@@ -0,0 +1,417 @@
+  switch (colDef.type) {
+    case 'number': {
+      return Number(value);
+    }
+    case 'boolean': {
+      return stringToBoolean(value);
+    }
+    case 'singleSelect': {
+      const colDefValueOptions = (colDef as GridSingleSelectColDef).valueOptions;
+      const valueOptions =
+        typeof colDefValueOptions === 'function'
+          ? colDefValueOptions({ field: colDef.field })
+          : colDefValueOptions || [];
+      const valueOption = valueOptions.find((option) => {
+        if (option === value) {
+          return true;
+        }
+        // TODO: would it work with valueFormatter?
+        if (typeof option === 'object' && option !== null) {
+          if (String(option.label) === value) {
+            return true;
+          }
+        }
+        return false;
+      });
+      if (valueOption) {
+        return valueOption;
+      }
+      return value;
+    }
+    case 'date':
+    case 'dateTime': {
+      const date = new Date(value);
+      return date;
+    }
+    default:
+      return value;
+  }
+};
+
+// Keeps track of updated rows during clipboard paste
+class CellValueUpdater {
+  rowsToUpdate: {
+    [rowId: GridRowId]: GridValidRowModel;
+  } = {};
+
+  updateRow: (row: GridRowModel) => void;
+
+  options: {
+    apiRef: React.MutableRefObject<GridPrivateApiPremium>;
+    processRowUpdate: DataGridPremiumProcessedProps['processRowUpdate'];
+    onProcessRowUpdateError: DataGridPremiumProcessedProps['onProcessRowUpdateError'];
+    getRowId: DataGridPremiumProcessedProps['getRowId'];
+  };
+
+  constructor(options: CellValueUpdater['options']) {
+    this.options = options;
+    this.updateRow = batchRowUpdates(options.apiRef.current.updateRows, 50);
+  }
+
+  updateCell({
+    rowId,
+    field,
+    pastedCellValue,
+  }: {
+    rowId: GridRowId;
+    field: GridColDef['field'];
+    pastedCellValue: string;
+  }) {
+    const { apiRef, getRowId } = this.options;
+    const colDef = apiRef.current.getColumn(field);
+    if (!colDef || !colDef.editable) {
+      return;
+    }
+    if (pastedCellValue === undefined) {
+      return;
+    }
+    const row = this.rowsToUpdate[rowId] || { ...apiRef.current.getRow(rowId) };
+    if (!row) {
+      return;
+    }
+
+    const parsedValue = parseCellStringValue(pastedCellValue, colDef);
+    if (parsedValue === undefined) {
+      return;
+    }
+
+    let rowCopy = { ...row };
+    if (typeof colDef.valueSetter === 'function') {
+      rowCopy = colDef.valueSetter({ value: parsedValue, row: rowCopy });
+    } else {
+      rowCopy[field] = parsedValue;
+    }
+    const newRowId = getRowIdFromRowModel(rowCopy, getRowId);
+    if (String(newRowId) !== String(rowId)) {
+      // We cannot update row id, so this cell value update should be ignored
+      return;
+    }
+    this.rowsToUpdate[rowId] = rowCopy;
+  }
+
+  applyUpdates() {
+    const { apiRef, processRowUpdate, onProcessRowUpdateError } = this.options;
+    const rowsToUpdate = this.rowsToUpdate;
```

Comment:
Avoid publishing event when no rows to update?
```suggestion
    const rowsToUpdate = this.rowsToUpdate;
    if (Object.keys(rowsToUpdate).length === 0) {
      return;
    }
```